---
title: More SQL
duration: "1:25"
creator:
    name: Francesco Mosconi
    city: SF
updated:
    name: David Yerrington
    city: SF
---

# ![](https://ga-dash.s3.amazonaws.com/production/assets/logo-9f88ae6c9c3871690e33280fcf557f33.png) More SQL
Week 5 | Lesson 3.1


**Please do not push this notebook to your personal Github repo so we can keep database credentials in-house ;)**

### LEARNING OBJECTIVES
*After this lesson, you will be able to:*
- Sort results by a column using `ORDER BY`
- Simplify our syntax using aliases (`AS`)
- Match patterns using `LIKE`
- Select distinct items using `DISTINCT`
- Aggregate values using `GROUP BY`
- Filter on aggregations using `HAVING`


### STUDENT PRE-WORK
*Before this lesson, you should already be able to:*
- connect to a local or remote relational database
- add and remove data, create new tables, alter table schemas
- perform simple queries using SQL language

### INSTRUCTOR PREP
*Before this lesson, instructors will need to:*
- Read in / Review any dataset(s) & starter/solution code
- Generate a brief slide deck
- Prepare any specific materials
- Provide students with additional resources
- make sure the GA copy of the northwind database is available at:
        psql -h dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com -p 5432 -U dsi_student northwind
        password: gastudents


### LESSON GUIDE / REFERENCE
| TIMING  | TYPE  | TOPIC  |
|:-:|---|---|
| 5 mins | [Opening](#opening) | Opening |
| 10 mins | [Introduction](#introduction) | More SQL |
| 10 mins | [Guided-Reference](#guided-practice_1) | Guided Reference: ORDER BY |
| 10 mins | [Guided-Reference](#guided-practice_2) | Guided Reference: Aliases |
| 10 mins | [Guided-Reference](#guided-practice_3) | Guided Reference: SQL LIKE Operator |
| 10 mins | [Guided-Reference](#guided-practice_4) | Guided Reference: DISTINCT and LIMIT Operators |
| 10 mins | [Guided-Reference](#guided-practice_5) | Guided Reference: GROUP BY Operator |
| 10 mins | [Guided-Reference](#guided-practice_6) | Guided Reference: HAVING Operator |
| 10 mins | [Ind-practice](#ind-practice) | Independent Practice |
| 5 mins | [Conclusion](#conclusion) | Conclusion |

{'/Users/davidyerrington/soundclips/soundclips_clean/a_team.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/alf.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/beatles.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/cheers.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/family_ties.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/friends_dionne.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/friends_whodini.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/ghostbusters.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/jeffersons.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/knight_rider.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/love_boat.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/magic_dance.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/magnificent.wav',
 '/Users/davidyerrington/soundclips/soundclips_clean/pokemon_1.wav',
 '/Users/davidyerrington

In [216]:
%%bash

afplay /Users/davidyerrington/soundclips/soundclips_clean/love_boat.wav

Process is interrupted.


# GROUP BY `table` Activity!

We are going to work together on the **Northwind** dataset to solidify our understanding of SQL basics.  We are going to split up into groups, and then we are going code for everyone like you wouldn't believe.

![](https://media.giphy.com/media/ZHlGzvZb130nm/giphy.gif)

Our activities will include:

- ** LIKE ** 
- ** GROUP** 
- ** DISTINCT and LIMIT**
- ** ORDER BY** 
- ** HAVING** (after we're comfortable with **`GROUP`** activities)

There will be a twist in this activity in which random person(s) will be selected from each group, each time, to present a solution on their given group topic, given a few minutes of deliberation and reference.  We will extend solutions based on actictivies frequenty so everyone has a chance to practice **and** discuss SQL implementation.

## First, Some Setup

In [217]:
# !pip install ipython-sql
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [218]:
# Special GROUP activity selector!
import random, glob
from subprocess import call

class activity:
    
    path       =  "/Users/davidyerrington/soundclips/soundclips_clean/"
    soundclips =  []
    
    def __init__(self):
        self.set_soundclips()
    
    def set_soundclips(self):

        soundclips = glob.glob(self.path + "*.wav")

        # Too provacative for class
        unsavory_clips = [ self.path + 'next_episode.wav', self.path + 'who_let_the_dogs_out.wav', self.path + 'welcome_mobb.wav' ]

        # Remove some bad ones because we wouldn't want swearing in class
        self.soundclips = list(set(soundclips) - set(unsavory_clips))
        print "set soundclips to: ", self.soundclips
    
    def select_group_activity(self, having=False, distinct=False, group=False, theme=True, student=False):

        # activities = ['LIKE', 'LIMIT', 'ORDER BY', 'ALIAS']
        activities = []

        if having:
            activities.append('HAVING')

        if group:
            activities.append('GROUP')

        if distinct:
            activities.append('DISTINCT')

        random_activity = random.choice(activities)

        text = "%s, your selected activity is. %s!" % (student, random_activity)

        print text
        call(["say", text])

        if theme:
            if len(self.soundclips) == 0:
                self.set_soundclips()
                
            clip = self.soundclips.pop()
            print "playing: ", clip, type(clip), len(clip)
            call(["afplay", clip])

In [295]:
from sqlalchemy import create_engine

# Because doesn't have `DESCRIBE [table]`
def describe_table(table="orders", connection=engine):
    
    sql = """
    SELECT
        column_name, data_type
    FROM
        information_schema.COLUMNS
    WHERE
        TABLE_NAME = '%s'
    """ % table

    return pd.read_sql(sql, con=connection)

### For Easy Reference

Here is a listing of tables in the **Northwind** database in our student Postgres database on **AWS**.

In [220]:
# You may need this package.  If so, run this in your `source activate dsi` environment
# conda install psycopg2

from sqlalchemy import create_engine
import pandas as pd
engine = create_engine('postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com:5432/northwind')


pd.read_sql("SELECT * FROM pg_catalog.pg_tables WHERE schemaname='public'", con=engine)

schemaname             tablename   tableowner tablespace hasindexes  \
0      public            categories          dsi       None       True   
1      public  customercustomerdemo          dsi       None       True   
2      public  customerdemographics          dsi       None       True   
3      public             customers          dsi       None       True   
4      public             employees          dsi       None       True   
5      public   employeeterritories          dsi       None       True   
6      public         order_details          dsi       None       True   
7      public                orders          dsi       None       True   
8      public              products          dsi       None       True   
9      public                region          dsi       None       True   
10     public              shippers          dsi       None       True   
11     public          shippers_tmp          dsi       None       True   
12     public             suppliers          dsi       None       True   
13     public           territories          dsi       None       True   
14     public              usstates          dsi       None      False   
15     public                table1  dsi_student       None      False   
16     public                movies  dsi_student       None       True   
17     public               ratings  dsi_student       None       True   

   hasrules hastriggers  
0     False       False  
1     False       False  
2     False       False  
3     False       False  
4     False       False  
5     False       False  
6     False       False  
7     False       False  
8     False       False  
9     False       False  
10    False       False  
11    False       False  
12    False       False  
13    False       False  
14    False       False  
15    False       False  
16    False       False  
17    False       False

## If you want to inspect the schema of a table, use this method

In [221]:
describe_table("orders", connection=engine)

column_name          data_type
0          OrderID           smallint
1       CustomerID          character
2       EmployeeID           smallint
3        OrderDate               date
4     RequiredDate               date
5      ShippedDate               date
6          ShipVia           smallint
7          Freight               real
8         ShipName  character varying
9      ShipAddress  character varying
10        ShipCity  character varying
11      ShipRegion  character varying
12  ShipPostalCode  character varying
13     ShipCountry  character varying

### Querying the database can be done 2 ways:

In [93]:
sql = """
SELECT * from territories LIMIT 4
"""

pd.read_sql(sql, con=engine)

TerritoryID TerritoryDescription  RegionID
0       01581             Westboro         1
1       01730              Bedford         1
2       01833            Georgetow         1
3       02116               Boston         1

## Inline with the sql notebook extension
_Sometimes this exentsion crashes, but trying it out again to see if the authors have fixed it up yet._

In [339]:
%%bash

afplay /Users/davidyerrington/soundclips/soundclips_clean/golden.mp3

Process is interrupted.


In [214]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
SELECT * FROM orders WHERE "RequiredDate" > '1996-07-01' LIMIT 10

KeyboardInterrupt: 

In [231]:
# Grab users signup
from StringIO import StringIO  # got moved to io in python3.
import pandas as pd

# import requests
# r = requests.get('https://docs.google.com/spreadsheets/d/1h31p7yWPZ5MKsTh3TJGQVIarKST3kX9bniRyNzdCSic/export?format=csv&id=1h31p7yWPZ5MKsTh3TJGQVIarKST3kX9bniRyNzdCSic&gid=0')
# data = r.content
# students = pd.read_csv(StringIO(data))
# # Student shuffle
# np.random.shuffle(students.values)
# students = students.values.tolist()

raw_students = """Ryan Dunlap
Chris
Denise Remhof
Lenny Levin 
Amos Kim
Alex K
Nicholle James
Daniel Moreno
Caitlin Mowdy
Kat Bishop
Ovid Jacob
Tim Lee
Nivetha Sivaprakasam
Kyle Mulle
Vanessa Grass
Cecilia
Divya
Alex H
Paul"""

students = raw_students.split("\n")
np.random.shuffle(students)
students

['Kat Bishop',
 'Caitlin Mowdy',
 'Alex H',
 'Ovid Jacob',
 'Tim Lee',
 'Alex K',
 'Divya',
 'Amos Kim',
 'Kyle Mulle',
 'Lenny Levin ',
 'Ryan Dunlap',
 'Vanessa Grass',
 'Paul',
 'Daniel Moreno',
 'Chris',
 'Nivetha Sivaprakasam',
 'Nicholle James',
 'Denise Remhof',
 'Cecilia']

In [232]:
# Initialize group activity class
dsi = activity()

set soundclips to:  ['/Users/davidyerrington/soundclips/soundclips_clean/pokemon_dave.wav', '/Users/davidyerrington/soundclips/soundclips_clean/taxi.wav', '/Users/davidyerrington/soundclips/soundclips_clean/knight_rider.wav', '/Users/davidyerrington/soundclips/soundclips_clean/jeffersons.wav', '/Users/davidyerrington/soundclips/soundclips_clean/russian.wav', '/Users/davidyerrington/soundclips/soundclips_clean/pokemon_1.wav', '/Users/davidyerrington/soundclips/soundclips_clean/alf.wav', '/Users/davidyerrington/soundclips/soundclips_clean/san_gabriel.wav', '/Users/davidyerrington/soundclips/soundclips_clean/family_ties.wav', '/Users/davidyerrington/soundclips/soundclips_clean/love_boat.wav', '/Users/davidyerrington/soundclips/soundclips_clean/war_power.wav', '/Users/davidyerrington/soundclips/soundclips_clean/transformers.wav', '/Users/davidyerrington/soundclips/soundclips_clean/magnificent.wav', '/Users/davidyerrington/soundclips/soundclips_clean/star_wars_cantina.wav', '/Users/davidyer

In [460]:
# Select random student, random activity, play their random theme song 
dsi.select_group_activity(student=students.pop(), having=True, distinct=True, group=True)
print "Students remaining: %d" % len(students)

# call(["afplay", "/Users/davidyerrington/soundclips/soundclips_clean/star_wars_battle.wav"])

Caitlin Mowdy, your selected activity is. DISTINCT!
playing:  /Users/davidyerrington/soundclips/soundclips_clean/alf.wav <type 'str'> 58


KeyboardInterrupt: 

<a name="opening"></a>
## Opening (5 mins)
We have seen how to connect to a local sqlite database and to a remote postgresql database.

**Check:** What SQL commands have we learned so far?

> Answer:
- CREATE
- INSERT
- DELETE
- UPDATE
- SELECT

**Check:** What different commands have we learned for SQLite and PostgreSQL?

> Answer:
- how to list schema and tables


## ContactName, first 20 from offset 10

In [296]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
SELECT "ContactName" FROM customers
LIMIT 20 OFFSET 10;

20 rows affected.


[(u'Victoria Ashworth',),
 (u'Patricio Simpson',),
 (u'Francisco Chang',),
 (u'Yang Wang',),
 (u'Pedro Afonso',),
 (u'Elizabeth Brown',),
 (u'Sven Ottlieb',),
 (u'Janine Labrune',),
 (u'Ann Devon',),
 (u'Roland Mendel',),
 (u'Aria Cruz',),
 (u'Diego Roel',),
 (u'Martine Ranc\xe9',),
 (u'Maria Larsson',),
 (u'Peter Franken',),
 (u'Carine Schmitt',),
 (u'Paolo Accorti',),
 (u'Lino Rodriguez',),
 (u'Eduardo Saavedra',),
 (u'Jos\xe9 Pedro Freyre',)]

## Frst 20 from Mexico sorted by company name - Chris

In [294]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
SELECT "CompanyName" FROM customers
WHERE "Country" LIKE 'Me%' LIMIT 20

5 rows affected.


[(u'Ana Trujillo Emparedados y helados',),
 (u'Antonio Moreno Taquer\xeda',),
 (u'Centro comercial Moctezuma',),
 (u'Pericles Comidas cl\xe1sicas',),
 (u'Tortuga Restaurante',)]

## Alias CompanyName, ContactName, ContactTitle from customers to something shorter

In [304]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
SELECT "CompanyName" as "compName", "ContactName" as "contName", "ContactTitle" as "contTitle" FROM customers
WHERE "ContactName" LIKE '%e';

4 rows affected.


[(u'Du monde entier', u'Janine Labrune', u'Owner'),
 (u'Godos Cocina T\xedpica', u'Jos\xe9 Pedro Freyre', u'Sales Manager'),
 (u'M\xe8re Paillarde', u'Jean Fresni\xe8re', u'Marketing Assistant'),
 (u'Wellington Importadora', u'Paula Parente', u'Sales Manager')]

## Show orders by ship date, order by ship date ascending, limit 10

In [311]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
SELECT * FROM "orders"
ORDER BY "ShippedDate" ASC
LIMIT 10

10 rows affected.


[(10249, u'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, u'Toms Spezialit\xe4ten', u'Luisenstr. 48', u'M\xfcnster', None, u'44087', u'Germany'),
 (10252, u'SUPRD', 4, datetime.date(1996, 7, 9), datetime.date(1996, 8, 6), datetime.date(1996, 7, 11), 2, 51.3, u'Supr\xeames d\xe9lices', u'Boulevard Tirou, 255', u'Charleroi', None, u'B-6000', u'Belgium'),
 (10250, u'HANAR', 4, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 12), 2, 65.83, u'Hanari Carnes', u'Rua do Pa\xe7o, 67', u'Rio de Janeiro', u'RJ', u'05454-876', u'Brazil'),
 (10251, u'VICTE', 3, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 15), 1, 41.34, u'Victuailles en stock', u'2, rue du Commerce', u'Lyon', None, u'69004', u'France'),
 (10255, u'RICSU', 9, datetime.date(1996, 7, 12), datetime.date(1996, 8, 9), datetime.date(1996, 7, 15), 3, 148.33, u'Richter Supermarkt', u'Starenweg 5', u'Gen\xe8ve', None, u'1204', u'Switzerland'),
 (10253, u'HANAR', 3, datetime.date(1996, 7, 10), datetime.date(1996, 7, 24), datetime.date(1996, 7, 16), 2, 58.17, u'Hanari Carnes', u'Rua do Pa\xe7o, 67', u'Rio de Janeiro', u'RJ', u'05454-876', u'Brazil'),
 (10248, u'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, u'Vins et alcools Chevalier', u"59 rue de l'Abbaye", u'Reims', None, u'51100', u'France'),
 (10256, u'WELLI', 3, datetime.date(1996, 7, 15), datetime.date(1996, 8, 12), datetime.date(1996, 7, 17), 2, 13.97, u'Wellington Importadora', u'Rua do Mercado, 12', u'Resende', u'SP', u'08737-363', u'Brazil'),
 (10257, u'HILAA', 4, datetime.date(1996, 7, 16), datetime.date(1996, 8, 13), datetime.date(1996, 7, 22), 3, 81.91, u'HILARION-Abastos', u'Carrera 22 con Ave. Carlos Soublette #8-35', u'San Crist\xf3bal', u'T\xe1chira', u'5022', u'Venezuela'),
 (10254, u'CHOPS', 5, datetime.date(1996, 7, 11), datetime.date(1996, 8, 8), datetime.date(1996, 7, 23), 2, 22.98, u'Chop-suey Chinese', u'Hauptstr. 31', u'Bern', None, u'3012', u'Switzerland')]

## SELECT order details, alias at least one column, order by Quantity AND UnitPrice

In [337]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
SELECT * , "Quantity" as quant FROM order_details
ORDER BY "Quantity" DESC, "UnitPrice" DESC
LIMIT 5; 

5 rows affected.


[(11072, 64, 33.25, 130, 0.0, 130),
 (10764, 39, 18.0, 130, 0.1, 130),
 (10776, 51, 53.0, 120, 0.05, 120),
 (10515, 27, 43.9, 120, 0.0, 120),
 (10711, 53, 32.8, 120, 0.0, 120)]

## Orders from countries starting with F and company names ending in E, ordered by dates DESC, from offset 3 limit 5

In [338]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
select * from orders where "ShipCountry" like 'F%' and "ShipName" like '%e' order by "ShippedDate" desc limit 5 offset 3 

5 rows affected.


[(10973, u'LACOR', 6, datetime.date(1998, 3, 24), datetime.date(1998, 4, 21), datetime.date(1998, 3, 27), 2, 15.17, u"La corne d'abondance", u"67, avenue de l'Europe", u'Versailles', None, u'78000', u'France'),
 (10972, u'LACOR', 4, datetime.date(1998, 3, 24), datetime.date(1998, 4, 21), datetime.date(1998, 3, 26), 2, 0.02, u"La corne d'abondance", u"67, avenue de l'Europe", u'Versailles', None, u'78000', u'France'),
 (10964, u'SPECD', 3, datetime.date(1998, 3, 20), datetime.date(1998, 4, 17), datetime.date(1998, 3, 24), 2, 87.38, u'Sp\xe9cialit\xe9s du monde', u'25, rue Lauriston', u'Paris', None, u'75016', u'France'),
 (10923, u'LAMAI', 7, datetime.date(1998, 3, 3), datetime.date(1998, 4, 14), datetime.date(1998, 3, 13), 3, 68.26, u"La maison d'Asie", u'1 rue Alsace-Lorraine', u'Toulouse', None, u'31000', u'France'),
 (10907, u'SPECD', 6, datetime.date(1998, 2, 25), datetime.date(1998, 3, 25), datetime.date(1998, 2, 27), 3, 9.19, u'Sp\xe9cialit\xe9s du monde', u'25, rue Lauriston', u'Paris', None, u'75016', u'France')]

## Find orders with shipping address containing "rue" or shipping address containing "z"

In [351]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
SELECT *
FROM "orders"
WHERE "ShipAddress" LIKE '%rue%' or "ShipAddress" LIKE '%z%'


118 rows affected.


[(10248, u'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, u'Vins et alcools Chevalier', u"59 rue de l'Abbaye", u'Reims', None, u'51100', u'France'),
 (10251, u'VICTE', 3, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 15), 1, 41.34, u'Victuailles en stock', u'2, rue du Commerce', u'Lyon', None, u'69004', u'France'),
 (10267, u'FRANK', 4, datetime.date(1996, 7, 29), datetime.date(1996, 8, 26), datetime.date(1996, 8, 6), 1, 208.58, u'Frankenversand', u'Berliner Platz 43', u'M\xfcnchen', None, u'80805', u'Germany'),
 (10274, u'VINET', 6, datetime.date(1996, 8, 6), datetime.date(1996, 9, 3), datetime.date(1996, 8, 16), 1, 6.01, u'Vins et alcools Chevalier', u"59 rue de l'Abbaye", u'Reims', None, u'51100', u'France'),
 (10276, u'TORTU', 8, datetime.date(1996, 8, 8), datetime.date(1996, 8, 22), datetime.date(1996, 8, 14), 3, 13.84, u'Tortuga Restaurante', u'Avda. Azteca 123', u'M\xe9xico D.F.', None, u'05033', u'Mexico'),
 (10279, u'LEHMS', 8, datetime.date(1996, 8, 13), datetime.date(1996, 9, 10), datetime.date(1996, 8, 16), 2, 25.83, u'Lehmanns Marktstand', u'Magazinweg 7', u'Frankfurt a.M.', None, u'60528', u'Germany'),
 (10284, u'LEHMS', 4, datetime.date(1996, 8, 19), datetime.date(1996, 9, 16), datetime.date(1996, 8, 27), 1, 76.56, u'Lehmanns Marktstand', u'Magazinweg 7', u'Frankfurt a.M.', None, u'60528', u'Germany'),
 (10293, u'TORTU', 1, datetime.date(1996, 8, 29), datetime.date(1996, 9, 26), datetime.date(1996, 9, 11), 3, 21.18, u'Tortuga Restaurante', u'Avda. Azteca 123', u'M\xe9xico D.F.', None, u'05033', u'Mexico'),
 (10295, u'VINET', 2, datetime.date(1996, 9, 2), datetime.date(1996, 9, 30), datetime.date(1996, 9, 10), 2, 1.15, u'Vins et alcools Chevalier', u"59 rue de l'Abbaye", u'Reims', None, u'51100', u'France'),
 (10304, u'TORTU', 1, datetime.date(1996, 9, 12), datetime.date(1996, 10, 10), datetime.date(1996, 9, 17), 2, 63.79, u'Tortuga Restaurante', u'Avda. Azteca 123', u'M\xe9xico D.F.', None, u'05033', u'Mexico'),
 (10311, u'DUMON', 1, datetime.date(1996, 9, 20), datetime.date(1996, 10, 4), datetime.date(1996, 9, 26), 3, 24.69, u'Du monde entier', u'67, rue des Cinquante Otages', u'Nantes', None, u'44000', u'France'),
 (10319, u'TORTU', 7, datetime.date(1996, 10, 2), datetime.date(1996, 10, 30), datetime.date(1996, 10, 11), 3, 64.5, u'Tortuga Restaurante', u'Avda. Azteca 123', u'M\xe9xico D.F.', None, u'05033', u'Mexico'),
 (10331, u'BONAP', 9, datetime.date(1996, 10, 16), datetime.date(1996, 11, 27), datetime.date(1996, 10, 21), 1, 10.19, u"Bon app'", u'12, rue des Bouchers', u'Marseille', None, u'13008', u'France'),
 (10332, u'MEREP', 3, datetime.date(1996, 10, 17), datetime.date(1996, 11, 28), datetime.date(1996, 10, 21), 2, 52.84, u'M\xe8re Paillarde', u'43 rue St. Laurent', u'Montr\xe9al', u'Qu\xe9bec', u'H1J 1C3', u'Canada'),
 (10334, u'VICTE', 8, datetime.date(1996, 10, 21), datetime.date(1996, 11, 18), datetime.date(1996, 10, 28), 2, 8.56, u'Victuailles en stock', u'2, rue du Commerce', u'Lyon', None, u'69004', u'France'),
 (10337, u'FRANK', 4, datetime.date(1996, 10, 24), datetime.date(1996, 11, 21), datetime.date(1996, 10, 29), 3, 108.26, u'Frankenversand', u'Berliner Platz 43', u'M\xfcnchen', None, u'80805', u'Germany'),
 (10339, u'MEREP', 2, datetime.date(1996, 10, 28), datetime.date(1996, 11, 25), datetime.date(1996, 11, 4), 2, 15.66, u'M\xe8re Paillarde', u'43 rue St. Laurent', u'Montr\xe9al', u'Qu\xe9bec', u'H1J 1C3', u'Canada'),
 (10340, u'BONAP', 1, datetime.date(1996, 10, 29), datetime.date(1996, 11, 26), datetime.date(1996, 11, 8), 3, 166.31, u"Bon app'", u'12, rue des Bouchers', u'Marseille', None, u'13008', u'France'),
 (10342, u'FRANK', 4, datetime.date(1996, 10, 30), datetime.date(1996, 11, 13), datetime.date(1996, 11, 4), 2, 54.83, u'Frankenversand', u'Berliner Platz 43', u'M\xfcnchen', None, u'80805', u'Germany'),
 (10343, u'LEHMS', 4, datetime.date(1996, 10, 31), datetime.date(1996, 11, 28), datetime.da

## Find orders by country having counts greater than 3

In [362]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
SELECT "ShipCountry", count(*)
FROM "orders"
group by "ShipCountry" having count(*) <= 10

2 rows affected.


[(u'Poland', 7L), (u'Norway', 6L)]

## how many customers by country?

In [368]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind

SELECT "Country", count("CustomerID")
FROM customers
group by "Country"
LIMIT 5

5 rows affected.


[(u'Austria', 2L),
 (u'Germany', 11L),
 (u'Venezuela', 4L),
 (u'Poland', 1L),
 (u'France', 11L)]

## Country with the most orders with highest "ShipVia" (order by largest)

In [383]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind

SELECT "ShipCountry", SUM("ShipVia") as qty 
FROM orders
GROUP BY "ShipCountry"
ORDER BY qty DESC
LIMIT 5

5 rows affected.


[(u'USA', 253L),
 (u'Germany', 231L),
 (u'Brazil', 152L),
 (u'France', 148L),
 (u'UK', 123L)]

## cities with more than 10 customers with phone numbers having "00" somewhere.. use an alias somwhere as well

In [407]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind

SELECT * FROM customers 
WHERE "Phone" LIKE '%00%'

3 rows affected.


[(u'ALFKI', u'Alfreds Futterkiste', u'Maria Anders', u'Sales Representative', u'Obere Str. 57', u'Berlin', None, u'12209', u'Germany', u'030-0074321', u'030-0076545'),
 (u'HANAR', u'Hanari Carnes', u'Mario Pontes', u'Accounting Manager', u'Rua do Pa\xe7o, 67', u'Rio de Janeiro', u'RJ', u'05454-876', u'Brazil', u'(21) 555-0091', u'(21) 555-8765'),
 (u'ROMEY', u'Romero y tomillo', u'Alejandra Camino', u'Accounting Manager', u'Gran V\xeda, 1', u'Madrid', None, u'28001', u'Spain', u'(91) 745 6200', u'(91) 745 6210')]

In [420]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
SELECT "City", count("CustomerID") FROM customers c
WHERE "Phone" LIKE '%1%'
GROUP BY "City"
ORDER BY count DESC

45 rows affected.


[(u'London', 6L),
 (u'Sao Paulo', 4L),
 (u'Buenos Aires', 3L),
 (u'Rio de Janeiro', 3L),
 (u'Madrid', 3L),
 (u'Lisboa', 2L),
 (u'Paris', 2L),
 (u'Charleroi', 1L),
 (u'Berlin', 1L),
 (u'Mannheim', 1L),
 (u'Kobenhavn', 1L),
 (u'Caracas', 1L),
 (u'Strasbourg', 1L),
 (u'Cunewalde', 1L),
 (u'Campinas', 1L),
 (u'I. de Margarita', 1L),
 (u'Gen\xe8ve', 1L),
 (u'Resende', 1L),
 (u'Warszawa', 1L),
 (u'Lille', 1L),
 (u'Toulouse', 1L),
 (u'San Crist\xf3bal', 1L),
 (u'Bruxelles', 1L),
 (u'Montr\xe9al', 1L),
 (u'Leipzig', 1L),
 (u'M\xfcnchen', 1L),
 (u'Stuttgart', 1L),
 (u'Aachen', 1L),
 (u'Lule\xe5', 1L),
 (u'Reims', 1L),
 (u'San Francisco', 1L),
 (u'Oulu', 1L),
 (u'Versailles', 1L),
 (u'Marseille', 1L),
 (u'Torino', 1L),
 (u'K\xf6ln', 1L),
 (u'Seattle', 1L),
 (u'Cowes', 1L),
 (u'Reggio Emilia', 1L),
 (u'Nantes', 1L),
 (u'Portland', 1L),
 (u'M\xfcnster', 1L),
 (u'Br\xe4cke', 1L),
 (u'Barquisimeto', 1L),
 (u'\xc5rhus', 1L)]

# Unique customer id's from customers, 2 ways

In [428]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
select DISTINCT "CustomerID" from customers;

select "CustomerID"
from customers
group by "CustomerID";

3 rows affected.


[(u'HashAggregate  (cost=3.14..4.05 rows=91 width=6)',),
 (u'  Group Key: "CustomerID"',),
 (u'  ->  Seq Scan on customers  (cost=0.00..2.91 rows=91 width=6)',)]

## Find order (from order details) where orders are over $10, count them

In [440]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
SELECT "OrderID", count(*)
FROM order_details
where ("UnitPrice" * "Quantity") > 10
group by "OrderID"


830 rows affected.


[(10501, 1L),
 (10295, 1L),
 (10827, 2L),
 (11051, 1L),
 (10423, 2L),
 (10945, 2L),
 (10527, 2L),
 (10529, 3L),
 (10470, 3L),
 (10645, 2L),
 (10495, 3L),
 (10468, 2L),
 (10406, 5L),
 (10266, 1L),
 (10657, 6L),
 (10890, 3L),
 (10551, 3L),
 (10884, 3L),
 (10616, 4L),
 (10911, 3L),
 (10510, 2L),
 (10738, 1L),
 (10893, 5L),
 (10430, 4L),
 (10780, 2L),
 (10792, 3L),
 (10916, 3L),
 (10394, 2L),
 (10459, 3L),
 (10761, 2L),
 (10543, 2L),
 (10647, 2L),
 (10627, 2L),
 (10694, 3L),
 (10342, 4L),
 (10323, 3L),
 (10427, 1L),
 (10484, 3L),
 (10396, 3L),
 (10798, 2L),
 (10990, 4L),
 (10324, 5L),
 (10434, 2L),
 (10514, 5L),
 (10710, 2L),
 (10767, 1L),
 (10711, 3L),
 (10726, 2L),
 (10876, 2L),
 (11068, 3L),
 (10253, 3L),
 (10579, 2L),
 (11062, 2L),
 (10517, 3L),
 (10351, 4L),
 (10379, 3L),
 (10558, 5L),
 (10838, 3L),
 (10595, 3L),
 (10982, 2L),
 (10264, 2L),
 (10298, 4L),
 (11003, 3L),
 (10498, 3L),
 (10846, 3L),
 (10882, 3L),
 (10633, 4L),
 (10970, 1L),
 (10349, 1L),
 (10329, 4L),
 (10826, 2L),
 (10917, 2L),
 (10675, 3L),
 (10822, 2L),
 (11017, 3L),
 (10978, 4L),
 (10322, 1L),
 (10965, 1L),
 (10939, 2L),
 (10766, 3L),
 (10956, 3L),
 (10912, 2L),
 (10428, 1L),
 (10600, 2L),
 (10383, 3L),
 (10463, 2L),
 (10770, 1L),
 (10995, 2L),
 (11001, 4L),
 (10900, 1L),
 (11039, 4L),
 (11049, 2L),
 (10784, 3L),
 (10401, 4L),
 (10471, 2L),
 (10515, 5L),
 (10333, 3L),
 (10691, 5L),
 (10299, 2L),
 (10462, 1L),
 (10814, 4L),
 (10614, 3L),
 (10618, 3L),
 (10474, 4L),
 (10268, 2L),
 (10796, 4L),
 (11073, 2L),
 (10779, 2L),
 (10829, 4L),
 (10875, 3L),
 (10986, 4L),
 (10331, 1L),
 (10773, 3L),
 (10440, 4L),
 (10823, 4L),
 (10715, 2L),
 (10317, 1L),
 (10562, 2L),
 (10456, 2L),
 (10685, 3L),
 (10281, 2L),
 (10910, 3L),
 (10743, 1L),
 (10388, 3L),
 (10594, 2L),
 (10870, 2L),
 (10603, 2L),
 (10320, 1L),
 (10411, 3L),
 (10420, 3L),
 (10426, 2L),
 (10721, 1L),
 (10759, 1L),
 (10480, 2L),
 (10378, 1L),
 (10297, 2L),
 (10490, 3L),
 (10717, 3L),
 (10927, 3L),
 (10500, 2L),
 (10395, 3L),
 (10727, 3L),
 (11002, 4L),
 (10387, 4L),
 (10719, 3L),
 (10696, 2L),
 (10904, 2L),
 (10418, 4L),
 (10573, 3L),
 (10380, 4L),
 (10530, 4L),
 (10748, 3L),
 (10692, 1L),
 (10940, 2L),
 (10526, 3L),
 (10293, 4L),
 (10628, 1L),
 (10442, 3L),
 (11030, 4L),
 (10698, 5L),
 (10860, 2L),
 (10385, 3L),
 (10376, 1L),
 (11014, 1L),
 (10905, 1L),
 (10550, 4L),
 (10554, 4L),
 (10563, 2L),
 (10254, 3L),
 (10304, 3L),
 (10853, 1L),
 (10817, 4L),
 (10467, 2L),
 (10834, 2L),
 (10781, 3L),
 (10416, 3L),
 (10439, 4L),
 (10683, 1L),
 (10811, 3L),
 (10485, 4L),
 (10847, 6L),
 (10923, 3L),
 (10273, 5L),
 (10311, 2L),
 (10932, 4L),
 (10895, 4L),
 (10334, 2L),
 (10382, 5L),
 (10607, 5L),
 (10896, 2L),
 (10279, 1L),
 (10355, 2L),
 (11033, 2L),
 (10262, 3L),
 (10622, 2L),
 (10857, 3L),
 (10753, 2L),
 (10632, 2L),
 (10699, 1L),
 (10522, 4L),
 (10914, 1L),
 (10638, 3L),
 (10733, 3L),
 (10867, 1L),
 (10954, 4L),
 (11050, 1L),
 (10248, 3L),
 (10520, 2L),
 (10575, 4L),
 (10801, 2L),
 (10454, 3L),
 (10409, 2L),
 (10436, 4L),
 (10693, 4L),
 (10513, 3L),
 (10994, 1L),
 (10449, 3L),
 (10690, 2L),
 (10613, 2L),
 (10372, 4L),
 (10414, 2L),
 (11057, 1L),
 (10926, 4L),
 (11028, 2L),
 (10650, 3L),
 (10667, 2L),
 (10547, 2L),
 (10836, 5L),
 (10286, 2L),
 (10477, 3L),
 (10531, 1L),
 (10588, 2L),
 (10277, 2L),
 (10356, 3L),
 (10589, 1L),
 (10635, 3L),
 (10673, 3L),
 (10570, 2L),
 (10634, 4L),
 (10604, 2L),
 (10855, 4L),
 (10996, 1L),
 (11036, 2L),
 (10902, 2L),
 (10270, 2L),
 (10864, 2L),
 (11011, 2L),
 (11018, 3L),
 (11055, 4L),
 (10948, 3L),
 (10422, 1L),
 (10483, 2L),
 (10519, 3L),
 (10854, 2L),
 (10695, 3L),
 (10966, 3L),
 (10360, 5L),
 (10736, 2L),
 (10944, 3L),
 (10820, 1L),
 (10924, 3L),
 (10339, 3L),
 (11040, 1L),
 (10260, 4L),
 (10429, 2L),
 (10481, 2L),
 (11004, 2L),
 (10818, 2L),
 (11010, 2L),
 (10363, 3L),
 (10476, 2L),
 (10641, 2L),
 (10296, 3L),
 (10508, 2L),
 (10671, 3L),
 (10751, 4L),
 (10291, 3L),
 (10469, 3L),
 (10561, 2L),
 (10578, 2L),
 (10707, 3L),
 (11076, 3L),
 (10371, 1L),
 (10466, 2L),
 (10934, 1L),
 (10313, 1

## unique customer ids from "orders" with employee ids > 10, in france, 

In [459]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
select "CustomerID", count(*)
from orders
where "EmployeeID" > 5 and "ShipCountry" = 'France'
group by "CustomerID"
order by count desc



9 rows affected.


[(u'LAMAI', 7L),
 (u'BONAP', 5L),
 (u'BLONP', 4L),
 (u'DUMON', 2L),
 (u'FOLIG', 2L),
 (u'VICTE', 2L),
 (u'SPECD', 1L),
 (u'LACOR', 1L),
 (u'VINET', 1L)]

## How many countries our orders by customerid span

In [465]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        

SELECT "CustomerID","ShipCountry", COUNT("ShipCountry")
FROM orders
group by "CustomerID", "ShipCountry"
        

89 rows affected.


[(u'BOTTM', u'Canada', 14L),
 (u'RICAR', u'Brazil', 11L),
 (u'FURIB', u'Portugal', 8L),
 (u'GROSR', u'Venezuela', 2L),
 (u'GODOS', u'Spain', 10L),
 (u'LEHMS', u'Germany', 15L),
 (u'MEREP', u'Canada', 13L),
 (u'SAVEA', u'USA', 31L),
 (u'NORTS', u'UK', 3L),
 (u'SUPRD', u'Belgium', 12L),
 (u'RANCH', u'Argentina', 5L),
 (u'CENTC', u'Mexico', 1L),
 (u'COMMI', u'Brazil', 5L),
 (u'TRADH', u'Brazil', 6L),
 (u'KOENE', u'Germany', 14L),
 (u'LAUGB', u'Canada', 3L),
 (u'VAFFE', u'Denmark', 11L),
 (u'REGGC', u'Italy', 12L),
 (u'TOMSP', u'Germany', 6L),
 (u'FRANK', u'Germany', 15L),
 (u'FOLIG', u'France', 5L),
 (u'RICSU', u'Switzerland', 10L),
 (u'WANDK', u'Germany', 10L),
 (u'QUEEN', u'Brazil', 13L),
 (u'LILAS', u'Venezuela', 14L),
 (u'WHITC', u'USA', 14L),
 (u'LACOR', u'France', 4L),
 (u'OTTIK', u'Germany', 10L),
 (u'LAZYK', u'USA', 2L),
 (u'MAISD', u'Belgium', 7L),
 (u'CONSH', u'UK', 3L),
 (u'LETSS', u'USA', 4L),
 (u'HANAR', u'Brazil', 14L),
 (u'LONEP', u'USA', 8L),
 (u'TRAIH', u'USA', 3L),
 (u'ANTON', u'Mexico', 7L),
 (u'FRANR', u'France', 3L),
 (u'BERGS', u'Sweden', 18L),
 (u'FRANS', u'Italy', 6L),
 (u'SPLIR', u'USA', 9L),
 (u'GOURL', u'Brazil', 9L),
 (u'ANATR', u'Mexico', 4L),
 (u'TORTU', u'Mexico', 10L),
 (u'EASTC', u'UK', 8L),
 (u'WOLZA', u'Poland', 7L),
 (u'BLAUS', u'Germany', 7L),
 (u'SANTG', u'Norway', 6L),
 (u'WILMK', u'Finland', 7L),
 (u'RATTC', u'USA', 18L),
 (u'FOLKO', u'Sweden', 19L),
 (u'PERIC', u'Mexico', 6L),
 (u'LAMAI', u'France', 14L),
 (u'WELLI', u'Brazil', 9L),
 (u'CHOPS', u'Switzerland', 8L),
 (u'WARTH', u'Finland', 15L),
 (u'MAGAA', u'Italy', 10L),
 (u'FAMIA', u'Brazil', 7L),
 (u'HUNGC', u'USA', 5L),
 (u'BSBEV', u'UK', 10L),
 (u'BOLID', u'Spain', 3L),
 (u'PRINI', u'Portugal', 5L),
 (u'PICCO', u'Austria', 10L),
 (u'BONAP', u'France', 17L),
 (u'HUNGO', u'Ireland', 19L),
 (u'ISLAT', u'UK', 10L),
 (u'ALFKI', u'Germany', 6L),
 (u'SPECD', u'France', 4L),
 (u'DRACD', u'Germany', 6L),
 (u'GALED', u'Spain', 5L),
 (u'HILAA', u'Venezuela', 18L),
 (u'ROMEY', u'Spain', 5L),
 (u'SIMOB', u'Denmark', 7L),
 (u'QUICK', u'Germany', 28L),
 (u'DUMON', u'France', 4L),
 (u'VINET', u'France', 5L),
 (u'OCEAN', u'Argentina', 5L),
 (u'MORGK', u'Germany', 5L),
 (u'ERNSH', u'Austria', 30L),
 (u'AROUT', u'UK', 13L),
 (u'THECR', u'USA', 3L),
 (u'BLONP', u'France', 11L),
 (u'GREAL', u'USA', 11L),
 (u'VICTE', u'France', 10L),
 (u'CACTU', u'Argentina', 6L),
 (u'OLDWO', u'USA', 10L),
 (u'QUEDE', u'Brazil', 9L),
 (u'THEBI', u'USA', 4L),
 (u'SEVES', u'UK', 9L),
 (u'LINOD', u'Venezuela', 12L)]

In [469]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
     
SELECT DISTINCT "CustomerID" FROM orders

89 rows affected.


[(u'HUNGO',),
 (u'VAFFE',),
 (u'PRINI',),
 (u'SEVES',),
 (u'TORTU',),
 (u'LONEP',),
 (u'LETSS',),
 (u'LAZYK',),
 (u'HILAA',),
 (u'WANDK',),
 (u'SPLIR',),
 (u'BOLID',),
 (u'FOLIG',),
 (u'FRANR',),
 (u'MAGAA',),
 (u'ISLAT',),
 (u'WHITC',),
 (u'QUEEN',),
 (u'BLAUS',),
 (u'ROMEY',),
 (u'RANCH',),
 (u'PICCO',),
 (u'CHOPS',),
 (u'CENTC',),
 (u'LINOD',),
 (u'TRADH',),
 (u'BERGS',),
 (u'FRANS',),
 (u'QUEDE',),
 (u'GODOS',),
 (u'SANTG',),
 (u'THEBI',),
 (u'OLDWO',),
 (u'LAUGB',),
 (u'FOLKO',),
 (u'CACTU',),
 (u'MEREP',),
 (u'FURIB',),
 (u'THECR',),
 (u'DRACD',),
 (u'BOTTM',),
 (u'SPECD',),
 (u'VINET',),
 (u'DUMON',),
 (u'RICSU',),
 (u'MORGK',),
 (u'WELLI',),
 (u'VICTE',),
 (u'SIMOB',),
 (u'NORTS',),
 (u'WARTH',),
 (u'HUNGC',),
 (u'LEHMS',),
 (u'PERIC',),
 (u'WILMK',),
 (u'RATTC',),
 (u'KOENE',),
 (u'ERNSH',),
 (u'TRAIH',),
 (u'TOMSP',),
 (u'FRANK',),
 (u'GALED',),
 (u'HANAR',),
 (u'ANATR',),
 (u'ANTON',),
 (u'GOURL',),
 (u'LACOR',),
 (u'AROUT',),
 (u'OTTIK',),
 (u'RICAR',),
 (u'LAMAI',),
 (u'BSBEV',),
 (u'COMMI',),
 (u'SAVEA',),
 (u'GROSR',),
 (u'OCEAN',),
 (u'GREAL',),
 (u'WOLZA',),
 (u'BONAP',),
 (u'MAISD',),
 (u'CONSH',),
 (u'ALFKI',),
 (u'LILAS',),
 (u'EASTC',),
 (u'QUICK',),
 (u'BLONP',),
 (u'REGGC',),
 (u'FAMIA',),
 (u'SUPRD',)]

<a name="introduction"></a>
## More SQL (10 mins)

In this lecture we'll learn a few more useful `SQL` commands that will give us the ability to perform more complex queries.

In particular we will learn to:
- Sort results by a column using `ORDER BY`
- Simplify our syntax using aliases
- Match patterns using `LIKE`
- Select distinct items using `DISTINCT`
- Aggregate values using `GROUP BY`
- Filter on aggregations using `HAVING`

For all the following example we will use the well-known [Northwind sample database](https://northwinddatabase.codeplex.com/).

We have imported this database into our PostgreSQL instance that you can find connecting here:

    psql -h dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com -p 5432 -U dsi_student northwind
    password: gastudents

We will use a few tables from this database, here is what they look like:

`customers`:

|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|ALFKI| Alfreds Futterkiste| Maria Anders | Sales Representative | Obere Str. 57 | Berlin|| 12209| Germany | 030-0074321| 030-0076545|
|ANATR| Ana Trujillo Emparedados y helados | Ana Trujillo | Owner| Avda. de la Constitución 2222 | México D.F. || 05021| Mexico| (5) 555-4729 | (5) 555-3745|
|ANTON| Antonio Moreno Taquería| Antonio Moreno | Owner| Mataderos2312 | México D.F. || 05023| Mexico| (5) 555-3932 |
|...|...|...|...|...|...|...|...|...|...|...|

`orders`:

|OrderID | CustomerID | EmployeeID | OrderDate| RequiredDate | ShippedDate | ShipVia | Freight | ShipName|ShipAddress |ShipCity| ShipRegion | ShipPostalCode | ShipCountry |
|----|
|10248 | VINET|5 | 1996-07-04 | 1996-08-01 | 1996-07-16| 3 | 32.38 | Vins et alcools Chevalier | 59 rue de l'Abbaye | Reims|| 51100| France|
|10249 | TOMSP|6 | 1996-07-05 | 1996-08-16 | 1996-07-10| 1 | 11.61 | Toms Spezialitäten| Luisenstr. 48| Münster|| 44087| Germany|
|10250 | HANAR|4 | 1996-07-08 | 1996-08-05 | 1996-07-12| 2 | 65.83 | Hanari Carnes | Rua do Paço, 67| Rio de Janeiro | RJ | 05454-876| Brazil|
|...|...|...|...|...|...|...|...|...|...|...|

`order_details`:

| OrderID |  ProductID |  UnitPrice | Quantity | Discount |
| ----- |
|10248|11|14|12|0|
|10248|42|9.8|10|0|
|10248|72|34.8|5|0|
|10249|14|18.6|9|0|
|10249|51|42.4|40|0|
|10250|41|7.7|10|0|
|...|...|...|...|...|


**Check** Let's come up with a few queries you'd be curious to run on these tables.


<a name="guided-practice_1"></a>
## Guided Practice: ORDER BY (10 mins)

The `ORDER BY` keyword is used to sort the result-set by one or more columns. It sorts the records in ascending order by default. To sort the records in a descending order, you can use the `DESC` keyword.

### SQL ORDER BY Syntax

```sql
SELECT _column_name_,_ column_name_  
FROM _table_name_  
ORDER BY _column_name _ASC|DESC,_ column_name_ ASC|DESC;
```

#### Example

The following SQL statement selects all customers from the "Customers" table,  sorted by the "Country" column:

```sql
SELECT "CustomerID", "CompanyName", "Country" FROM customers
ORDER BY "Country";
```

|CustomerID|CompanyName|Country|
|----|
|OCEAN|Océano Atlántico Ltda.|Argentina|
|CACTU|Cactus Comidas para llevar|Argentina|
|RANCH|Rancho grande |Argentina|
|...|...|...|


#### ORDER BY DESC Example

The following SQL statement selects all customers from the "Customers" table,  sorted DESCENDING by the "Country" column:


```sql
SELECT "CustomerID", "CompanyName", "Country" FROM customers
ORDER BY "Country" DESC;
```

|CustomerID|CompanyName|Country|
|----|
|GROSR|GROSELLA-Restaurante |Venezuela|
|LILAS|LILA-Supermercado|Venezuela|
|HILAA|HILARION-Abastos |Venezuela|
|...|...|...|



#### ORDER BY Several Columns Example

The following SQL statement selects all customers from the `customers` table,  sorted by the "Country" and the "CustomerName" column:

```sql
SELECT "CustomerID", "CompanyName", "Country" FROM customers
ORDER BY "Country", "CompanyName";
```

|CustomerID|CompanyName|Country|
|----|
|CACTU|Cactus Comidas para llevar|Argentina|
|OCEAN|Océano Atlántico Ltda.|Argentina|
|RANCH|Rancho grande |Argentina|
|...|...|...|

**Check** run a few queries on the provided database and try using the `ORDER BY` command.

<a name="guided-practice_2"></a>
## Guided Practice: Aliases (10 mins)

SQL aliases are used to give a database table, or a column in a table, a temporary name. Basically aliases are created to make column names more readable.

### SQL Alias Syntax for Columns

```sql
SELECT _column_name_ AS _alias_name_  
FROM _table_name;_
```
### SQL Alias Syntax for Tables
```sql
SELECT _column_name(s)_  
FROM _table_name _AS _alias_name;_
```

#### Alias Examples for Table Columns

The following SQL statement specifies two aliases, one for the `CompanyName`  column and one for the `ContactName` column:

```sql
SELECT "CompanyName" AS "Customer", "ContactName" AS "[Contact Person]"
FROM customers;
```

|Customer|[Contact Person]|
|----|
|Alfreds Futterkiste| Maria Anders|
|Ana Trujillo Emparedados y helados| Ana Trujillo|
|Antonio Moreno Taquería| Antonio Moreno|
|...|...|


In the following SQL statement we combine four columns (Address, City, PostalCode and Country) and create an alias named `Address`:

```sql
SELECT "CompanyName" AS "Customer",
    CONCAT("Address",', ',"City",', ', "PostalCode",', ',"Country") AS "Address"
FROM customers;
```

|Customer|Address|
|----|
|Alfreds Futterkiste| Obere Str. 57, Berlin, 12209, Germany|
|Ana Trujillo Emparedados y helados| Avda. de la Constitución 2222, México D.F., 05021, Mexico|
|Antonio Moreno Taquería Mataderos| 2312, México D.F., 05023, Mexico|
|...|...|


#### Alias Example for Tables

The following SQL statement selects all the orders from the customer with  CustomerID=4 (Around the Horn). We use the `customers` and `orders` tables, and  give them the table aliases of `c` and `o` respectively (Here we have used  aliases to make the SQL shorter):


```sql
SELECT o."OrderID", o."OrderDate", c."CompanyName"
FROM customers AS c, orders  AS o  
WHERE c."CompanyName" = 'Around the Horn' AND c."CustomerID"=o."CustomerID";
```

|CustomerID|OrderDate|CompanyName|
|----|
|10355|1996-11-15|Around the Horn|
|10383|1996-12-16|Around the Horn|
|10453|1997-02-21|Around the Horn|
|10558|1997-06-04|Around the Horn|
|...|...|...|

**Check** what would be the same SQL statement without aliases?

Aliases can be useful when:

- There are more than one table involved in a query
- Functions are used in the query
- Column names are big or not very readable
- Two or more columns are combined together

**Check** try running the same queries as above with aliases. Are they more readable?

<a name="guided-practice_3"></a>
## Guided Practice: SQL LIKE Operator (10 mins)

The LIKE operator is used in a WHERE clause to search for a specified pattern in a column.

### SQL LIKE Syntax
```sql
SELECT _column_name(s)_  
FROM _table_name_  
WHERE _column_name_ LIKE _pattern_;
```

#### SQL LIKE Operator Examples

The following SQL statement selects all customers with a City starting with  the letter "S":

```sql
SELECT * FROM customers
WHERE "City" LIKE 'S%';
```

|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|BLONP| Blondesddsl père et fils|Frédérique Citeaux|Marketing Manager| 24, place Kléber|Strasbourg||67000| France|88.60.15.31|88.60.15.32|
|COMMI| Comércio Mineiro|Pedro Afonso|Sales Associate|Av. dos Lusíadas, 23|Sao Paulo| SP|05432-043| Brazil|(11) 555-7647| 
|FAMIA| Familia Arquibaldo|Aria Cruz| Marketing Assistant | Rua Orós, 92|Sao Paulo| SP|05442-030| Brazil|(11) 555-9857| 
|...|...|...|...|...|...|...|...|...|...|...|

**Tip:** The "%" sign is used to define wildcards (missing  letters) both before and after the pattern. Also notice that `PostgreSQL` is case sensitive.

The following SQL statement selects all customers with a City ending with the  letter "s":

```sql
SELECT * FROM customers
WHERE "City" LIKE '%s';
```

|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|CACTU| Cactus Comidas para llevar|Patricio Simpson|Sales Agent|Cerrito 333 |Buenos Aires||1010|Argentina| (1) 135-5555|(1) 135-4892|
|DUMON| Du monde entier|Janine Labrune|Owner| 67, rue des Cinquante Otages|Nantes||44000| France|40.67.88.88| 40.67.89.89|
|FRANR| France restauration|Carine Schmitt|Marketing Manager| 54, rue Royale|Nantes||44000| France|40.32.21.21| 40.32.21.20|
|...|...|...|...|...|...|...|...|...|...|...|

Using the NOT keyword allows you to select records that do NOT match the pattern. The following SQL statement selects all customers with Country NOT  containing the pattern "land":

```sql
SELECT * FROM customers
WHERE "Country" NOT LIKE '%land%';
```
|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|ALFKI| Alfreds Futterkiste|Maria Anders|Sales Representative|Obere Str. 57| Berlin||12209| Germany|030-0074321| 030-0076545|
|ANATR| Ana Trujillo Emparedados y helados|Ana Trujillo|Owner| Avda. de la Constitución 2222| México D.F.|| 05021| Mexico|(5) 555-4729|(5) 555-3745|
|ANTON| Antonio Moreno Taquería|Antonio Moreno|Owner| Mataderos|2312 México D.F.|| 05023| Mexico|(5) 555-3932|
|...|...|...|...|...|...|...|...|...|...|...|

**Check** run a few queries on the provided database and try using the `LIKE` command.

<a name="guided-practice_4"></a>
## Guided Practice: DISTINCT and LIMIT Operators (10 mins)

The SELECT DISTINCT statement is used to return only distinct (different) values. In a table, a column may contain many duplicate values; and sometimes you only want to list the different (distinct) values.

### SQL SELECT DISTINCT Syntax
```sql
SELECT DISTINCT _column_name_,_column_name_  
FROM _table_name_;
```

#### SELECT DISTINCT Example

The following SQL statement selects only the distinct values from the "City"  columns from the "Customers" table:

```sql
SELECT DISTINCT "City" FROM customers;
```

| City|
|---|
|Leipzig|
|London|
|Nantes|
|...|

The `DISTINCT` operator is equivalent to the `numpy.unique` command.

### SQL SELECT LIMIT Syntax
Sometimes we may want to only retrieve a fixed number of records from the database. This is where the `LIMIT` operator comes in.

```sql
SELECT _column_name_,_column_name_  
FROM _table_name_
LIMIT _number_of_records;
```

#### SELECT LIMIT Example

The following SQL statement selects only the first 3 values from the "City"  columns from the "Customers" table:

```sql
SELECT DISTINCT "City"
FROM customers
LIMIT 3;
```

| City|
|---|
|Leipzig|
|London|
|Nantes|

The `LIMIT` operator is equivalent to the `pandas.Dataframe.head` command.

**Check** try running the same queries using `DISTINCT` and `LIMIT`

<a name="guided-practice_5"></a>
## Guided Practice: GROUP BY Operator (10 mins)

A table may contain several records that have a common key. Consider for example the following `order_details` table:


| OrderID |  ProductID |  UnitPrice | Quantity | Discount |
| ----- |
|10248|11|14|12|0|
|10248|42|9.8|10|0|
|10248|72|34.8|5|0|
|10249|14|18.6|9|0|
|10249|51|42.4|40|0|
|10250|41|7.7|10|0|
|...|...|...|...|...|


The _GROUP BY_ statement is used in conjunction with the aggregate functions to group the result-set by one or more columns. For example we may want to know the total number of items purchased in each order.

#### SQL GROUP BY Syntax

The syntax to perform a _GROUP BY_ operation is the following:

```sql
SELECT column_name, aggregate_function(column_name)  
FROM table_name  
WHERE column_name operator value  
GROUP BY column_name;
```
We select the column we want to aggregate on and the function.


#### Example

In order to calculate the total amount of items in a specific order we can use the following query:

```sql
SELECT "OrderID", SUM("Quantity")
FROM order_details
GROUP BY "OrderID"
```

Note that in `PostgreSQL` unquoted names are case-insensitive. Thus `SELECT * FROM hello` and `SELECT * FROM HELLO` are equivalent.

However, quoted names are case-sensivite. `SELECT * FROM "hello"` is not equivalent to `SELECT * FROM "HELLO"`.

To make a "bridge" between quoted names and unquoted names, unquoted names are implicitly lowercased, thus hello, HELLO and HeLLo are equivalent to "hello", but not to "HELLO" or "HeLLo".

Thus, when creating entities (tables, views, procedures, etc) in PostgreSQL, you should specify them either unquoted, or quoted-but-lowercased.

<a name="guided-practice_6"></a>
## Guided Practice: HAVING Operator (10 mins)

The HAVING clause was added to SQL because the WHERE keyword could not be used with aggregate functions. Let's consider the previous example where we aggregated the `Quantity` of an order. What if we would like to filter the results for orders where the Maximum discout applied on any item is lower than 10%?

#### SQL HAVING Syntax

```sql
SELECT column_name, aggregate_function(column_name)  
FROM table_name  
WHERE column_name operator value  
GROUP BY column_name  
HAVING aggregate_function(column_name) operator value;
```

##### Example

```sql
SELECT "OrderID", SUM("Quantity"), MAX("Discount")
FROM order_details
GROUP BY "OrderID"
HAVING MAX("Discount") <= 0.1;
```

|OrderID|sum|max|
|---|
|10501|20|0|
|10295|4|0|
|10827|36|0|

<a name="ind-practice"></a>
## Independent Practice (10 minutes)

Now that you've learned a lot more commands, try combining them:

- retrieve the top 3 orders by number of items ordered
- retrieve the oldest 5 orders
- retrieve the 5 most recent orders. Nest that query into another one that ranks these 5 orders by shipping country (alphabetical order)
- try some queries on your own

> Answers:
```sql
SELECT "OrderID", SUM("Quantity") AS "Sum"
FROM order_details
GROUP BY "OrderID"
ORDER BY "Sum" DESC
LIMIT 3;
```
```sql
SELECT "OrderID", "OrderDate"
FROM orders
ORDER BY "OrderDate"
LIMIT 5;
```
```sql
SELECT * FROM (
	SELECT "OrderID", "OrderDate", "ShipCountry"
	FROM orders
	ORDER BY "OrderDate" DESC
	LIMIT 5
	) AS fivemostrecent
ORDER BY "ShipCountry";
```

<a name="conclusion"></a>
## Conclusion (5 mins)
In this lesson we have learned many more new commands to make our SQL queries more powerful.


In particular we have learned how to:
- Sort results by a column using `ORDER BY`
- Simplify our syntax using aliases
- Match patterns using `LIKE`
- Select distinct items using `DISTINCT`
- Aggregate values using `GROUP BY`
- Filter on aggregations using `HAVING`

**Check** can you think of a few more business cases where these are useful?

***

### ADDITIONAL RESOURCES

- [PostgreSQL Documentation](http://www.postgresql.org/docs/)
- [Sqlite Documentation](https://www.sqlite.org/docs.html)